core/controller/src/main/scala/whisk/core/controller/actions/PrimitiveActions.scala

```diff
@@ -139,6 +199,356 @@ protected[actions] trait PrimitiveActions {
+   * @param action the conductor action to invoke for the composition
+   * @param payload the dynamic arguments for the activation
+   * @param waitForResponse if not empty, wait upto specified duration for a response (this is used for blocking activations)
+   * @param cause the activation id that is responsible for this invoke/activation
+   * @param caller the session object for the caller if any
+   * @param transid a transaction id for logging
+   * @return a promise that completes with one of the following successful cases:
+   *            Right(WhiskActivation) if waiting for a response and response is ready within allowed duration,
+   *            Left(ActivationId) if not waiting for a response, or allowed duration has elapsed without a result ready
+   *         or an error
+   */
+  private def invokeComposition(
+    user: Identity,
+    action: ExecutableWhiskActionMetaData,
+    payload: Option[JsObject],
+    waitForResponse: Option[FiniteDuration],
+    cause: Option[ActivationId],
+    caller: Option[Session] = None)(implicit transid: TransactionId): Future[Either[ActivationId, WhiskActivation]] = {
+
+    val session = Session(
+      activationId = activationIdFactory.make(),
+      start = Instant.now(Clock.systemUTC()),
+      action,
+      cause,
+      duration = 0,
+      maxMemory = action.limits.memory.megabytes,
+      state = None,
+      accounting = caller.map { _.accounting }.getOrElse(CompositionAccounting()), // share accounting with caller
+      logs = Buffer.empty,
+      caller,
+      result = waitForResponse.map { _ =>
+        Promise[Either[ActivationId, WhiskActivation]]() // placeholder for result if blocking invoke
+      })
+
+    invokeConductor(user, payload, session)
+
+    // is caller waiting for the result of the activation?
+    waitForResponse
+      .map { timeout =>
+        // handle timeout
+        session.result.head.future
+          .withAlternativeAfterTimeout(timeout, Future.successful(Left(session.activationId)))
+      }
+      .getOrElse {
+        // no, return the session id
+        Future.successful(Left(session.activationId))
+      }
+  }
+
+  /**
+   * A method that knows how to handle a conductor invocation.
+   *
+   * This method prepares the payload and invokes the conductor action.
+   * It parses the result and extracts the name of the next component action if any.
+   * It either invokes the desired component action or completes the composition invocation.
+   * It also checks the invocation counts against the limits.
+   *
+   * @param user the identity invoking the action
+   * @param payload the dynamic arguments for the activation
+   * @param session the session object for this composition
+   * @param transid a transaction id for logging
+   */
+  private def invokeConductor(user: Identity, payload: Option[JsObject], session: Session)(
+    implicit transid: TransactionId) = {
+
+    if (session.accounting.conductors > 2 * actionSequenceLimit) {
```

Comment:
I don't understand what you mean by static value. The maximum session length is configurable using the same configuration variable as the maximum sequence length. The limit cannot be enforced statically as the length of the composition is not known statically. The factor 2 here is somewhat arbitrary and chosen so that the limit on composition length is primarily enforced by counting the number of composed actions (as opposed to conductor actions) but to make sure nevertheless that no infinite recursion is possible. See later test in code: `if (session.accounting.components >= actionSequenceLimit)`.

Reply:
I mean lift the multiplication out to a static field. Separately we might want a different configuration parameter for continuations vs sequences to remove the 2x.

Reply:
I have two tests: `if (session.accounting.conductors > 2 * actionSequenceLimit)` and `if (session.accounting.components >= actionSequenceLimit)`. Should these two limits by independent additional configuration parameters?